emcc.py

```diff
@@ -2235,27 +2251,31 @@ def check_memory_setting(setting):
     if settings.GLOBAL_BASE != -1:
       exit_with_error("ASan does not support custom GLOBAL_BASE")
 
-    max_mem = settings.INITIAL_MEMORY
+    user_mem = settings.INITIAL_MEMORY
     if settings.ALLOW_MEMORY_GROWTH:
-      max_mem = settings.MAXIMUM_MEMORY
+      user_mem = settings.MAXIMUM_MEMORY
 
-    shadow_size = max_mem // 8
-    settings.GLOBAL_BASE = shadow_size
+    # The total memory is 1/8th shadow and 7/8th user/usable memory.
+    # Given the know value of user memory size we can work backwards
+    # to find the total memory and the shadow size based on the fact
+    # that the user memory is 7/8ths of the total memory.
+    # (i.e. user_mem == total_mem * 7 / 8
+    total_mem = user_mem * 8 / 7
```

Comment:
I am slightly worried about rounding errors here, since I assume ASan rounds up, that is, isn't the size of shadow memory potentially slightly more than user memory / 7?

I would compute this is
```python
shadow_mem = ..however asan rounds..
total_mem = user_mem + shadow_mem
```

Or does ASan actually compute it as a fraction of total memory, not user memory? In that case your code lgtm

Reply:
ASan literally does a shift go from a pointer to its shadow byte:

https://github.com/emscripten-core/emscripten/blob/f195d7cde910a9fee876b4c02e012f03e1e85c18/system/lib/compiler-rt/lib/asan/asan_mapping_emscripten.h#L46

The problem with trying to start by calculating the size of the shadow region is that you can't do that without also knowing the size of the shadow region.. since the bigger the shadow region the more shadow region you need because it shift the reset of the memory upwards as it grows.

Reply:
The assertion that I've added to the code I think means that if these calculation are wrong we would know right away..   at least if we ever under-estimate the size of the shadow region which is the dangerous direction.

Reply:
Wait if `SHADOW_SCALE=3` then total memory is 9/8 times user memory, isn't it, and not 8/7?

Reply:
Because `SHADOW_SCALE=3` the size of the shadow memory is 1/8th size of the total memory.    But shadow memory is also part of the total memory and because shadow memory lives at address zero that leave 7/8ths of the total memory left, for what we are calling user memory.

Reply:
For what its worth the calculation of `__global_base` in emcc.py is puts its *exactly* at the address calculated for `max_shadow_address` in the native code..  which seems like fairly good evidence that the calculation here are correct.

Reply:
Ah, right. sgtm